File: abstract-factory.go
// Certainly! Below is a complex example of the Abstract Factory design pattern in Go. This example demonstrates the creation of different types of computers with varying components, such as processors and memory:

// Copy
package main

import "fmt"

// AbstractProduct interfaces
type Processor interface {
	Specification() string
}

type Memory interface {
	Specification() string
}

// ConcreteProduct classes
type HighEndProcessor struct{}

func (p *HighEndProcessor) Specification() string {
	return "High-end Processor"
}

type BasicProcessor struct{}

func (p *BasicProcessor) Specification() string {
	return "Basic Processor"
}

type LargeMemory struct{}

func (m *LargeMemory) Specification() string {
	return "Large Memory"
}

type SmallMemory struct{}

func (m *SmallMemory) Specification() string {
	return "Small Memory"
}

// AbstractFactory interface
type ComputerFactory interface {
	CreateProcessor() Processor
	CreateMemory() Memory
}

// ConcreteFactory classes
type HighEndComputerFactory struct{}

func (f *HighEndComputerFactory) CreateProcessor() Processor {
	return &HighEndProcessor{}
}

func (f *HighEndComputerFactory) CreateMemory() Memory {
	return &LargeMemory{}
}

type BasicComputerFactory struct{}

func (f *BasicComputerFactory) CreateProcessor() Processor {
	return &BasicProcessor{}
}

func (f *BasicComputerFactory) CreateMemory() Memory {
	return &SmallMemory{}
}

// Client code
func main() {
	// Creating a high-end computer
	highEndFactory := &HighEndComputerFactory{}
	highEndProcessor := highEndFactory.CreateProcessor()
	highEndMemory := highEndFactory.CreateMemory()

	fmt.Println("High-end Computer:")
	fmt.Println("Processor: " + highEndProcessor.Specification())
	fmt.Println("Memory: " + highEndMemory.Specification())

	fmt.Println()

	// Creating a basic computer
	basicFactory := &BasicComputerFactory{}
	basicProcessor := basicFactory.CreateProcessor()
	basicMemory := basicFactory.CreateMemory()

	fmt.Println("Basic Computer:")
	fmt.Println("Processor: " + basicProcessor.Specification())
	fmt.Println("Memory: " + basicMemory.Specification())
}
In this example:

Processor and Memory are abstract product interfaces representing different components of a computer.
HighEndProcessor, BasicProcessor, LargeMemory, and SmallMemory are concrete product classes implementing the respective product interfaces.
ComputerFactory is the abstract factory interface declaring methods to create Processor and Memory.
HighEndComputerFactory and BasicComputerFactory are concrete factory classes implementing the ComputerFactory interface, providing different implementations of CreateProcessor and CreateMemory to create families of related products.
The client code demonstrates how different factories create families of products, and it uses these products to build computers.
This example showcases the Abstract Factory pattern in Go, allowing you to create families of related objects with varying components while ensuring that the created objects are compatible with each other.

File: abstract-factory.java
// Certainly! The Abstract Factory design pattern provides an interface for creating families of related or dependent objects without specifying their concrete classes. Here's a complex example of the Abstract Factory pattern in Java:

// Copy
// Abstract product interfaces
interface Engine {
    String getSpecification();
}

interface Tires {
    String getSpecification();
}

// Concrete product classes
class LuxuryEngine implements Engine {
    @Override
    public String getSpecification() {
        return "Luxury Engine";
    }
}

class StandardEngine implements Engine {
    @Override
    public String getSpecification() {
        return "Standard Engine";
    }
}

class ExpensiveTires implements Tires {
    @Override
    public String getSpecification() {
        return "Expensive Tires";
    }
}

class CheapTires implements Tires {
    @Override
    public String getSpecification() {
        return "Cheap Tires";
    }
}

// Abstract factory interface
interface CarFactory {
    Engine createEngine();
    Tires createTires();
}

// Concrete factory classes
class LuxuryCarFactory implements CarFactory {
    @Override
    public Engine createEngine() {
        return new LuxuryEngine();
    }

    @Override
    public Tires createTires() {
        return new ExpensiveTires();
    }
}

class StandardCarFactory implements CarFactory {
    @Override
    public Engine createEngine() {
        return new StandardEngine();
    }

    @Override
    public Tires createTires() {
        return new CheapTires();
    }
}

// Client code
public class AbstractFactoryPatternExample {
    public static void main(String[] args) {
        // Creating a luxury car
        CarFactory luxuryFactory = new LuxuryCarFactory();
        Engine luxuryEngine = luxuryFactory.createEngine();
        Tires luxuryTires = luxuryFactory.createTires();

        System.out.println("Luxury Car:");
        System.out.println("Engine: " + luxuryEngine.getSpecification());
        System.out.println("Tires: " + luxuryTires.getSpecification());

        System.out.println();

        // Creating a standard car
        CarFactory standardFactory = new StandardCarFactory();
        Engine standardEngine = standardFactory.createEngine();
        Tires standardTires = standardFactory.createTires();

        System.out.println("Standard Car:");
        System.out.println("Engine: " + standardEngine.getSpecification());
        System.out.println("Tires: " + standardTires.getSpecification());
    }
}
In this example:

Engine and Tires are abstract product interfaces representing different components of a car.
LuxuryEngine, StandardEngine, ExpensiveTires, and CheapTires are concrete product classes implementing the respective product interfaces.
CarFactory is the abstract factory interface declaring methods to create Engine and Tires.
LuxuryCarFactory and StandardCarFactory are concrete factory classes implementing the CarFactory interface, providing different implementations of createEngine and createTires to create families of related products.
The client code demonstrates how different factories create families of products, and it uses these products to build cars.
The Abstract Factory pattern allows you to create families of related objects (in this case, different components of a car) without specifying their concrete classes. This promotes flexibility and ensures that the created objects are compatible with each other.